File: typescript/src/parsers/genericParser.ts
/**
 * Generic parsers for Terraform blocks.
 * Handles resource, data, provider, module, terraform settings, and unknown block types.
 */

import {
    DataBlock,
    GenericBlock,
    ModuleBlock,
    ProviderBlock,
    ResourceBlock,
    TerraformSettingsBlock
} from '../types/blocks';
import { HclBlock } from '../types/blocks';
import { parseBlockBody } from '../utils/parser/bodyParser';
import { literalString } from '../utils/common/valueHelpers';

/** Meta-argument keys that are separated from resource properties */
const META_KEYS = new Set(['count', 'for_each', 'provider', 'depends_on', 'lifecycle']);

/**
 * Parser for terraform settings blocks.
 *
 * @example
 * ```hcl
 * terraform {
 *   required_version = ">= 1.0.0"
 *   required_providers {
 *     aws = {
 *       source  = "hashicorp/aws"
 *       version = "~> 4.0"
 *     }
 *   }
 * }
 * ```
 */
export class TerraformSettingsParser {
    /**
     * Parses a terraform settings block.
     * @param block - The raw HCL block to parse
     * @returns Parsed TerraformSettingsBlock
     */
    parse(block: HclBlock): TerraformSettingsBlock {
        const parsed = parseBlockBody(block.body);
        return {
            properties: parsed.attributes,
            raw: block.raw,
            source: block.source
        };
    }
}

/**
 * Parser for provider configuration blocks.
 *
 * @example
 * ```hcl
 * provider "aws" {
 *   alias  = "west"
 *   region = "us-west-2"
 * }
 * ```
 */
export class ProviderParser {
    /**
     * Parses a provider configuration block.
     * @param block - The raw HCL block to parse
     * @returns Parsed ProviderBlock with name, alias, and properties
     */
    parse(block: HclBlock): ProviderBlock {
        const name = block.labels[0] || 'default';
        const parsed = parseBlockBody(block.body);

        return {
            name,
            alias: literalString(parsed.attributes.alias) ?? parsed.attributes.alias?.raw,
            properties: parsed.attributes,
            raw: block.raw,
            source: block.source
        };
    }
}

/**
 * Parser for module call blocks.
 *
 * @example
 * ```hcl
 * module "vpc" {
 *   source  = "terraform-aws-modules/vpc/aws"
 *   version = "3.0.0"
 *
 *   name = "my-vpc"
 *   cidr = "10.0.0.0/16"
 * }
 * ```
 */
export class ModuleParser {
    /**
     * Parses a module call block.
     * @param block - The raw HCL block to parse
     * @returns Parsed ModuleBlock with name and all properties
     */
    parse(block: HclBlock): ModuleBlock {
        const name = block.labels[0] || 'unnamed';
        const parsed = parseBlockBody(block.body);

        return {
            name,
            properties: parsed.attributes,
            raw: block.raw,
            source: block.source
        };
    }
}

/**
 * Parser for resource definition blocks.
 * Separates meta-arguments (count, for_each, depends_on, etc.) from resource properties.
 *
 * @example
 * ```hcl
 * resource "aws_instance" "web" {
 *   count         = 3
 *   ami           = var.ami_id
 *   instance_type = "t2.micro"
 *
 *   tags = {
 *     Name = "web-${count.index}"
 *   }
 *
 *   dynamic "ebs_block_device" {
 *     for_each = var.ebs_volumes
 *     content {
 *       device_name = ebs_block_device.value.device_name
 *       volume_size = ebs_block_device.value.size
 *     }
 *   }
 * }
 * ```
 */
export class ResourceParser {
    /**
     * Parses a resource block, separating properties, meta-arguments, and dynamic blocks.
     * @param block - The raw HCL block to parse
     * @returns Parsed ResourceBlock with separated concerns
     */
    parse(block: HclBlock): ResourceBlock {
        const [type, name] = block.labels;
        const parsed = parseBlockBody(block.body);
        const meta: ResourceBlock['meta'] = {};
        const properties: ResourceBlock['properties'] = {};

        // Separate meta-arguments from properties
        for (const [key, value] of Object.entries(parsed.attributes)) {
            if (META_KEYS.has(key)) {
                meta[key] = value;
            } else {
                properties[key] = value;
            }
        }

        return {
            type: type || 'unknown',
            name: name || 'unnamed',
            properties,
            blocks: parsed.blocks.filter((child) => child.type !== 'dynamic'),
            dynamic_blocks: this.extractDynamicBlocks(parsed.blocks),
            meta,
            raw: block.raw,
            source: block.source
        };
    }

    /**
     * Extracts dynamic blocks from nested blocks.
     * @param blocks - Nested blocks from the resource body
     * @returns Array of DynamicBlock objects
     */
    private extractDynamicBlocks(blocks: ReturnType<typeof parseBlockBody>['blocks']): ResourceBlock['dynamic_blocks'] {
        const dynamicBlocks: ResourceBlock['dynamic_blocks'] = [];

        for (const block of blocks) {
            if (block.type !== 'dynamic') {
                continue;
            }

            const label = block.labels[0] || 'dynamic';
            const for_each = block.attributes.for_each;
            const iterator = literalString(block.attributes.iterator);
            const contentBlock = block.blocks.find((child) => child.type === 'content');

            dynamicBlocks.push({
                label,
                for_each,
                iterator,
                content: contentBlock?.attributes || {},
                raw: block.raw
            });
        }

        return dynamicBlocks;
    }
}

/**
 * Parser for data source blocks.
 *
 * @example
 * ```hcl
 * data "aws_ami" "ubuntu" {
 *   most_recent = true
 *
 *   filter {
 *     name   = "name"
 *     values = ["ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-*"]
 *   }
 *
 *   owners = ["099720109477"]
 * }
 * ```
 */
export class DataParser {
    /**
     * Parses a data source block.
     * @param block - The raw HCL block to parse
     * @returns Parsed DataBlock with type, name, properties, and nested blocks
     */
    parse(block: HclBlock): DataBlock {
        const [dataType, name] = block.labels;
        const parsed = parseBlockBody(block.body);

        return {
            dataType: dataType || 'unknown',
            name: name || 'unnamed',
            properties: parsed.attributes,
            blocks: parsed.blocks,
            raw: block.raw,
            source: block.source
        };
    }
}

/**
 * Parser for generic/unknown block types.
 * Used for moved, import, check, terraform_data, and unrecognized blocks.
 *
 * @example
 * ```hcl
 * moved {
 *   from = aws_instance.old
 *   to   = aws_instance.new
 * }
 *
 * import {
 *   to = aws_instance.example
 *   id = "i-1234567890abcdef0"
 * }
 *
 * check "health" {
 *   assert {
 *     condition     = data.http.health.status_code == 200
 *     error_message = "Health check failed"
 *   }
 * }
 * ```
 */
export class GenericBlockParser {
    /**
     * Parses a generic block into a GenericBlock structure.
     * @param block - The raw HCL block to parse
     * @returns Parsed GenericBlock with type, labels, properties, and nested blocks
     */
    parse(block: HclBlock): GenericBlock {
        const parsed = parseBlockBody(block.body);
        return {
            type: block.keyword,
            labels: block.labels,
            properties: parsed.attributes,
            blocks: parsed.blocks,
            raw: block.raw,
            source: block.source
        };
    }
}

File: python/src/parse_hcl/parsers/generic_parser.py
"""
Generic parsers for Terraform block types.

Provides parsers for terraform settings, provider, module, resource, data,
and generic blocks.
"""

from __future__ import annotations

from typing import Dict, List

from ..types import HclBlock, Value
from ..utils.common.value_helpers import literal_string
from ..utils.parser.body_parser import parse_block_body

META_KEYS = {"count", "for_each", "provider", "depends_on", "lifecycle"}


class TerraformSettingsParser:
    """Parser for terraform settings blocks."""

    def parse(self, block: HclBlock) -> Dict[str, object]:
        """
        Parses a terraform settings block.

        Args:
            block: The raw HCL block.

        Returns:
            Parsed terraform settings with properties.
        """
        parsed = parse_block_body(block["body"])
        return {
            "properties": parsed["attributes"],
            "raw": block["raw"],
            "source": block["source"],
        }


class ProviderParser:
    """Parser for provider configuration blocks."""

    def parse(self, block: HclBlock) -> Dict[str, object]:
        """
        Parses a provider configuration block.

        Args:
            block: The raw HCL block.

        Returns:
            Parsed provider with name, alias, and properties.
        """
        labels = block.get("labels", [])
        name = labels[0] if labels else "default"
        parsed = parse_block_body(block["body"])
        alias_value = parsed["attributes"].get("alias")
        alias = literal_string(alias_value) or (alias_value.get("raw") if isinstance(alias_value, dict) else None)

        return {
            "name": name,
            "alias": alias,
            "properties": parsed["attributes"],
            "raw": block["raw"],
            "source": block["source"],
        }


class ModuleParser:
    """Parser for module call blocks."""

    def parse(self, block: HclBlock) -> Dict[str, object]:
        """
        Parses a module call block.

        Args:
            block: The raw HCL block.

        Returns:
            Parsed module with name and properties.
        """
        labels = block.get("labels", [])
        name = labels[0] if labels else "unnamed"
        parsed = parse_block_body(block["body"])
        return {
            "name": name,
            "properties": parsed["attributes"],
            "raw": block["raw"],
            "source": block["source"],
        }


class ResourceParser:
    """Parser for resource definition blocks."""

    def parse(self, block: HclBlock) -> Dict[str, object]:
        """
        Parses a resource definition block.

        Extracts resource type, name, properties, nested blocks, dynamic blocks,
        and meta-arguments.

        Args:
            block: The raw HCL block.

        Returns:
            Parsed resource with type, name, properties, blocks, and meta.
        """
        labels = block.get("labels", [])
        # Safe label extraction with fallbacks
        resource_type = labels[0] if len(labels) > 0 else "unknown"
        name = labels[1] if len(labels) > 1 else "unnamed"
        parsed = parse_block_body(block["body"])

        meta: Dict[str, Value] = {}
        properties: Dict[str, Value] = {}

        for key, value in parsed["attributes"].items():
            if key in META_KEYS:
                meta[key] = value
            else:
                properties[key] = value

        return {
            "type": resource_type,
            "name": name,
            "properties": properties,
            "blocks": [b for b in parsed["blocks"] if b.get("type") != "dynamic"],
            "dynamic_blocks": self._extract_dynamic_blocks(parsed["blocks"]),
            "meta": meta,
            "raw": block["raw"],
            "source": block["source"],
        }

    def _extract_dynamic_blocks(self, blocks: List[Dict[str, object]]) -> List[Dict[str, object]]:
        """
        Extracts dynamic block definitions from nested blocks.

        Args:
            blocks: List of nested blocks.

        Returns:
            List of parsed dynamic block structures.
        """
        dynamic_blocks: List[Dict[str, object]] = []

        for child in blocks:
            if child.get("type") != "dynamic":
                continue

            labels = child.get("labels")
            if not isinstance(labels, list):
                labels = []
            label = labels[0] if labels else "dynamic"

            attributes = child.get("attributes", {})
            if not isinstance(attributes, dict):
                attributes = {}

            iterator_value = attributes.get("iterator")
            iterator = literal_string(iterator_value) if isinstance(iterator_value, dict) else None

            content_block = None
            nested_blocks = child.get("blocks", [])
            if isinstance(nested_blocks, list):
                for nested in nested_blocks:
                    if isinstance(nested, dict) and nested.get("type") == "content":
                        content_block = nested
                        break

            dynamic_blocks.append(
                {
                    "label": label,
                    "for_each": attributes.get("for_each"),
                    "iterator": iterator,
                    "content": content_block.get("attributes", {}) if content_block else {},
                    "raw": child.get("raw"),
                }
            )

        return dynamic_blocks


class DataParser:
    """Parser for data source definition blocks."""

    def parse(self, block: HclBlock) -> Dict[str, object]:
        """
        Parses a data source definition block.

        Args:
            block: The raw HCL block.

        Returns:
            Parsed data source with dataType, name, properties, and blocks.
        """
        labels = block.get("labels", [])
        # Safe label extraction with fallbacks
        data_type = labels[0] if len(labels) > 0 else "unknown"
        name = labels[1] if len(labels) > 1 else "unnamed"
        parsed = parse_block_body(block["body"])

        return {
            "dataType": data_type,
            "name": name,
            "properties": parsed["attributes"],
            "blocks": parsed["blocks"],
            "raw": block["raw"],
            "source": block["source"],
        }


class GenericBlockParser:
    """Parser for generic/unknown block types."""

    def parse(self, block: HclBlock) -> Dict[str, object]:
        """
        Parses a generic block (moved, import, check, etc.).

        Args:
            block: The raw HCL block.

        Returns:
            Parsed block with type, labels, properties, and nested blocks.
        """
        parsed = parse_block_body(block["body"])
        return {
            "type": block["keyword"],
            "labels": block.get("labels", []),
            "properties": parsed["attributes"],
            "blocks": parsed["blocks"],
            "raw": block["raw"],
            "source": block["source"],
        }
